src/base/unicode_utils.cpp

```diff
@@ -138,9 +147,13 @@ std::wstring utf8_to_ucs2(const std::string& str8) {
 }
 #else
 std::string ucs2_to_utf8(const std::wstring& str16) {
+  return ucs2_to_utf8(str16.c_str(), str16.c_str() + str16.length());
+}
+
+std::string ucs2_to_utf8(const wchar_t* begin, const wchar_t* end) {
   std::string result;
-  const auto* cursor = str16.c_str();
-  const auto* const end = str16.c_str() + str16.length();
+  result.reserve(static_cast<size_t>(std::max(static_cast<int>(end - begin), 0)));
```

Comment:
I think you could do with a single cast (*and* support >2 GB strings ;-) ):

```C++
  result.reserve(std::max(static_cast<size_t>(end - begin), 0U));
```

Reply:
I've made two casts to prevent possible crash, in case incorrect `end` passed to the function: `end < begin`, it's kinda over engineering, so I'm gonna remove both